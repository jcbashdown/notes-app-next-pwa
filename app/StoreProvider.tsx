'use client'
import { useRef } from 'react'
import { Provider } from 'react-redux'
import { makeStore, AppStore } from '@/lib/redux/store'
import { initFromRxDB, initFromFixture, setNoteTopic, setCursorPosition } from '@/lib/redux/features/noteSlice'

export default function StoreProvider({ children, fixture = null }: { children: React.ReactNode; fixture: any }) {
    const storeRef = useRef<AppStore | null>(null)

    if (!storeRef.current) {
        // Create the store instance the first time this renders
        storeRef.current = makeStore()
        //sync the store with rxdb
        if (fixture) {
            storeRef.current
                .dispatch(initFromFixture(fixture))
                .unwrap()
                .then((initializationData) => {
                    const { noteTopics } = initializationData
                    //TODO - remove this in future - user must select manually. Using
                    //it now so we don't have to implement selection yet
                    //if (noteTopics.length > 0) {
                    //storeRef.current!.dispatch(setNoteTopic(noteTopics[0].id))
                    //}
                })
        } else {
            storeRef.current
                .dispatch(initFromRxDB())
                .unwrap()
                .then((initializationData) => {
                    const { noteTopics } = initializationData
                    //TODO - remove this in future - user must select manually. Using
                    //it now so we don't have to implement selection yet
                    //if (noteTopics.length > 0) {
                    //storeRef.current!.dispatch(setNoteTopic(noteTopics[0].id))
                    //}
                })
        }
    }

    return <Provider store={storeRef.current}>{children}</Provider>
}
