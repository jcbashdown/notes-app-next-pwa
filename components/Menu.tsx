'use client'

import { useState, useEffect, useRef } from 'react'
import { useAppSelector } from '@/lib/redux/hooks'
import { selectNotesById, selectNoteChildrenByParentId } from '@/lib/redux/features/noteSlice'
import NewTopic from '@/components/NewTopic'
import generateNotesJSONFromStoreState from '@/lib/redux/utils/generateNotesJSONFromStoreState'
import { IdNoteRelationsMap } from '@/lib/redux/features/noteSlice'
import { NoteDocType } from '@/lib/rxdb/types/noteTypes'
export default function Menu() {
    const [isDropdownVisible, setIsDropdownVisible] = useState(false)
    const dropdownRef = useRef<HTMLDivElement | null>(null)

    const noteChildrenByParentId = useAppSelector(selectNoteChildrenByParentId)
    const notes = Object.values(useAppSelector(selectNotesById))

    useEffect(() => {
        document.addEventListener('mousedown', handleClickOutside)

        return () => {
            document.removeEventListener('mousedown', handleClickOutside)
        }
    }, [])
    const toggleDropdown = () => {
        setIsDropdownVisible(!isDropdownVisible)
    }

    const handleClickOutside = (event: MouseEvent) => {
        console.log(event.target, dropdownRef.current)
        if (dropdownRef.current && !dropdownRef.current.contains(event.target as Node)) {
            setIsDropdownVisible(false)
        }
    }
    const handleClick = () => {
        console.log('downloading')
        downloadJSON(notes, noteChildrenByParentId)
    }
    const downloadJSON = (
        notes: NoteDocType[],
        noteChildrenByParentId: IdNoteRelationsMap,
        filename = 'notes.json'
    ) => {
        console.log('downloading')
        const json = generateNotesJSONFromStoreState(notes, noteChildrenByParentId)
        const blob = new Blob([json], { type: 'application/json' })
        const href = URL.createObjectURL(blob)
        const link = document.createElement('a')
        link.href = href
        link.download = filename
        document.body.appendChild(link)
        link.click()
        document.body.removeChild(link)
    }
    return (
        <div className="bg-white p-4 rounded-lg shadow-md mb-4 flex">
            <NewTopic />
            <div ref={dropdownRef}>
                <button
                    onClick={toggleDropdown}
                    className="ml-2 p-2 border border-gray-300 rounded-lg focus:outline-none focus:ring-2 focus:ring-blue-400"
                    id="menuButton"
                >
                    <svg
                        className="w-6 h-6 text-gray-700"
                        fill="none"
                        stroke="currentColor"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                    >
                        <path
                            strokeLinecap="round"
                            strokeLinejoin="round"
                            strokeWidth="2"
                            d="M4 6h16M4 12h16M4 18h16"
                        ></path>
                    </svg>
                </button>
                {/*TODO - style this better*/}
                <div
                    className={`${
                        isDropdownVisible ? 'block' : 'hidden'
                    } absolute right-0 mt-2 w-48 bg-white border border-gray-200 rounded-lg shadow-lg`}
                    id="menuDropdown"
                >
                    <a
                        id="download-note-json"
                        href="#"
                        onClick={handleClick}
                        className="block px-4 py-2 text-gray-800 hover:bg-gray-200"
                    >
                        Download notes
                    </a>
                </div>
            </div>
        </div>
    )
}
