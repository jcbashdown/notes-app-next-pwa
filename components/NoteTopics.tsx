import NoteTopic from '@/components/NoteTopic'

import { useAppSelector } from '@/lib/redux/hooks'
import { selectNoteTopics } from '@/lib/redux/features/noteSlice'

export default function NoteTopics() {
    const noteTopics = useAppSelector(selectNoteTopics)

    return (
        <>
            {noteTopics.map((noteTopic, i) => (
                <NoteTopic key={i} noteTopic={noteTopic} />
            ))}
        </>
    )
}
