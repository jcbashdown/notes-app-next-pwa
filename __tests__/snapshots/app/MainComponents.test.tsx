import { render, act } from '@testing-library/react'
import e2eTestFixture from '@/fixtures/e2e-test-fixture.json'
import StoreProvider from '@/app/StoreProvider'
import Menu from '@/components/Menu'
import NoteTopics from '@/components/NoteTopics'

export default function MainComponentsWithFixtureFromE2ETests() {
    return (
        <>
            {/* Menu Section */}
            <Menu />
            {/* Notes List */}
            <div className="space-y-4">
                <NoteTopics />
            </div>
        </>
    )
}

it('renders homepage with e2e fixtures unchanged', async () => {
    const { container } = await act(async () =>
        render(
            <StoreProvider fixture={e2eTestFixture}>
                <MainComponentsWithFixtureFromE2ETests />
            </StoreProvider>
        )
    )
    expect(container).toMatchSnapshot()
})
