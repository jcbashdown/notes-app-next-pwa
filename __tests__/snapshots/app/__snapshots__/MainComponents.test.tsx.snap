// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders homepage with e2e fixtures unchanged 1`] = `
<div>
  <div
    class="bg-white p-4 rounded-lg shadow-md mb-4 flex"
  >
    <input
      class="w-full p-2 border border-gray-300 rounded-lg focus:outline-none focus:ring-2 focus:ring-blue-400"
      placeholder="Add a new note..."
      type="text"
      value=""
    />
    <div>
      <button
        class="ml-2 p-2 border border-gray-300 rounded-lg focus:outline-none focus:ring-2 focus:ring-blue-400"
        id="menuButton"
      >
        <svg
          class="w-6 h-6 text-gray-700"
          fill="none"
          stroke="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 6h16M4 12h16M4 18h16"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
      <div
        class="hidden absolute right-0 mt-2 w-48 bg-white border border-gray-200 rounded-lg shadow-lg"
        id="menuDropdown"
      >
        <a
          class="block px-4 py-2 text-gray-800 hover:bg-gray-200"
          href="#"
          id="download-note-json"
        >
          Download notes
        </a>
      </div>
    </div>
  </div>
  <div
    class="space-y-4"
  >
    <div
      class="bg-white p-4 rounded-lg shadow-md"
    >
      <input
        type="text"
        value="NewnotNewnote"
      />
      <ul
        class="ml-2"
      >
        <li>
          <span
            class="p-1"
          >
            <input
              class="pl-1"
              type="text"
              value="+"
            />
            <input
              type="text"
              value="This note supports the Newnote topic"
            />
          </span>
          <ul
            class="ml-2"
          />
        </li>
        <li>
          <span
            class="p-1"
          >
            <input
              class="pl-1"
              type="text"
              value="-"
            />
            <input
              type="text"
              value="This not opposes the Newnote topic"
            />
          </span>
          <ul
            class="ml-2"
          />
        </li>
        <li>
          <span
            class="p-1"
          >
            <input
              class="pl-1"
              type="text"
              value=""
            />
            <input
              type="text"
              value="This note is related"
            />
          </span>
          <ul
            class="ml-2"
          >
            <li>
              <span
                class="p-1"
              >
                <input
                  class="pl-1"
                  type="text"
                  value="+"
                />
                <input
                  type="text"
                  value="This note supports the related note"
                />
              </span>
              <ul
                class="ml-2"
              />
            </li>
          </ul>
        </li>
      </ul>
    </div>
    <div
      class="bg-white p-4 rounded-lg shadow-md"
    >
      <input
        type="text"
        value="Helloagain!"
      />
    </div>
    <div
      class="bg-white p-4 rounded-lg shadow-md"
    >
      <input
        type="text"
        value="This is the root note"
      />
    </div>
  </div>
</div>
`;
